### PR TITLE
feat: adding support for creating targets from ExternalName

### PR DESCRIPTION
**Pull request by msvticket**
### Issue

 #3505

### Description

I added support for ip target groups when service is of type ExternalName.

I have not updated any documentation, but I'm happy to do that in case you want the code change to begin with.

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by msvticket** [comment]
Could you have a look at this @oliviassss ?

**Comment by Leen15** [comment]
any update on this?

**Comment by Reyalsorik** [comment]
Any update? I would appreciate this functionality.

**Comment by nousot-cloud-guy** [comment]
Is there anything that's holding up this PR that someone could help finish? This would be really useful functionality to have.

**Comment by felipegouveiae** [comment]
@johngmyers @M00nF1sh please review and merge! I'm begging you!

File: pkg/backend/endpoint_resolver.go [hunk 1]
```diff
@@ -27,12 +28,17 @@ var ErrNotFound = errors.New("backend not found")
 type EndpointResolver interface {
 	// ResolvePodEndpoints will resolve endpoints backed by pods directly.
 	// returns resolved podEndpoints and whether there are unready endpoints that can potentially turn ready in future reconciles.
-	ResolvePodEndpoints(ctx context.Context, svcKey types.NamespacedName, port intstr.IntOrString,
-		opts ...EndpointResolveOption) ([]PodEndpoint, bool, error)
+	ResolvePodEndpoints(ctx context.Context, svckey types.NamespacedName, svc *corev1.Service, port intstr.IntOrString, opts ...EndpointResolveOption) ([]IpEndpoint, bool, error)
```
Comment (ryantate13) [inline pkg/backend/endpoint_resolver.go:31]: There's no need for a breaking API change here. The variadic list of `EndpointResolveOption` argument should eliminate that necessity, can just create a new one to the effect of `WithService(*corev1.Service)`. This package is currently at v2, merging this would require tagging a v3 and that creates more churn for adoption than necessary.

File: pkg/backend/endpoint_types.go [hunk 2]
```diff
@@ -12,18 +12,18 @@ type Endpoint interface {
 	GetIdentifier(includeTimestamp bool) string
 }
 
-// An endpoint provided by pod directly.
-type PodEndpoint struct {
+// IpEndpoint is an endpoint for an ip address
+type IpEndpoint struct {
```
Comment (ryantate13) [inline pkg/backend/endpoint_types.go:16]: Name here should remain stable

File: pkg/backend/endpoint_types.go [hunk 3]
```diff
@@ -12,18 +12,18 @@ type Endpoint interface {
...
 	// Pod's IP.
 	IP string
 	// Pod's container port.
 	Port int32
 	// Pod that provides this endpoint.
-	Pod k8s.PodInfo
+	Pod *k8s.PodInfo
```
Comment (ryantate13) [inline pkg/backend/endpoint_types.go:22]: This needs to remain backwards compatible

**Review by ryantate13 — CHANGES_REQUESTED** [review]
I'm excited about this functionality getting built. I do think we need to be cognizant of dependents on this package and maintain a stable API surface for v2 however. Seems like some simple fix ups should get is back into that territory.

**Comment by msvticket** [comment]
> I'm excited about this functionality getting built. I do think we need to be cognizant of dependents on this package and maintain a stable API surface for v2 however. Seems like some simple fix ups should get is back into that territory.

I don't mind putting some more time in this PR if it could be merged. But since this PR is ignored by the maintainers I don't see any point in spending that time now.

File: pkg/backend/endpoint_types.go (hunk 2 above)
Comment (msvticket) [reply pkg/backend/endpoint_types.go:16]: Why?

File: pkg/backend/endpoint_types.go (hunk 3 above)
Comment (msvticket) [reply pkg/backend/endpoint_types.go:22]: Why?

**Comment by deens-cyera** [comment]
Hey, any update in the pr ?

**Comment by chris-codaio** [comment]
Another vote to get this reviewed & committed - just ran into this limitation this week!

File: pkg/ingress/model_build_target_group.go [hunk 4]
```diff
@@ -228,12 +228,16 @@ func (t *defaultModelBuildTask) buildTargetGroupName(_ context.Context,
...
-func (t *defaultModelBuildTask) buildTargetGroupTargetType(_ context.Context, svcAndIngAnnotations map[string]string, classCfg ClassConfiguration) (elbv2model.TargetType, error) {
+func (t *defaultModelBuildTask) buildTargetGroupTargetType(_ context.Context, svcAndIngAnnotations map[string]string, classCfg ClassConfiguration, svc *corev1.Service) (elbv2model.TargetType, error) {
 	rawTargetType := string(t.defaultTargetType)
 	_ = t.annotationParser.ParseStringAnnotation(annotations.IngressSuffixTargetType, &rawTargetType, svcAndIngAnnotations)
 	if classCfg.IngClassParams != nil && classCfg.IngClassParams.Spec.TargetType != "" {
 		rawTargetType = string(classCfg.IngClassParams.Spec.TargetType)
 	}
+	if svc.Spec.Type == corev1.ServiceTypeExternalName && rawTargetType != string(elbv2model.TargetTypeIP) {
+		t.logger.Info("Target type will be ip since service is an ExternalName", "service", k8s.NamespacedName(svc))
+		rawTargetType = string(elbv2model.TargetTypeIP)
```
Comment (zac-nixon) [inline pkg/ingress/model_build_target_group.go:266]: I'd recommend throwing an error rather then overwriting this value.

Comment (msvticket) [reply pkg/ingress/model_build_target_group.go:266]: I think returning an error if something else than ip is set in an annotation makes sense, but still default to ip. What do you think about that?

Comment (msvticket) [reply pkg/ingress/model_build_target_group.go:266]: I have now pushed a commit with that change

**Comment by zac-nixon** [comment]
Hi. I'd like to update everyone on the conversation we had in Slack.

```
Zac Nixon
  [1:55 PM](https://kubernetes.slack.com/archives/D09ARVB64TF/p1755723355218379)
I reviewed the PR, the code your wrote makes sense but I have a concern. How does the controller know that the DNS record has been updated, so that it can update the target group? Right now, you're relying the sync time (~10 hours) to do the DNS updates. It seems that we would need some background process that is querying dns periodically and enqueuing events when the record changes.


Mårten Svantesson
  [8:51 AM](https://kubernetes.slack.com/archives/D09ARVB64TF/p1755791499599059)
Well, my thought is that this is an initial implementation that I tried to keep as simple as possible. You are right in that it probably wont suffice for everybody, but I think it is better to base improvements on feedback rather than guesses.


Zac Nixon
  [7:53 AM](https://kubernetes.slack.com/archives/D09ARVB64TF/p1755874427108159)
Yeah, I see what you're saying. I really appreciate the PR, but generally we don't accept features that aren't complete. Not having a way to timely update the LoadBalancer when the DNS entries changes seems like a big problem. Can you explain how this feature helps you out?


Mårten Svantesson
  [7:09 AM](https://kubernetes.slack.com/archives/D09ARVB64TF/p1756130961627789)
I use it to implement this solution: https://aws.amazon.com/blogs/networking-and-content-delivery/hosting-internal-https-static-websites-with-alb-s3-and-privatelink/
By using an ExternalName for s3 I don’t need a separate ALB but can use the same that is used for accessing internal application running in EKS.
Amazon Web ServicesAmazon Web Services
[Hosting Internal HTTPS Static Websites with ALB, S3, and PrivateLink | Amazon Web Services](https://aws.amazon.com/blogs/networking-and-content-delivery/hosting-internal-https-static-websites-with-alb-s3-and-privatelink/)
Amazon Simple Storage Service (Amazon S3) is a powerful platform that enables you to do various tasks. One notable feature is the ability to create a bucket with an FQDN, point an alias record to the bucket website endpoint, and immediately get up-and-running with an HTTP static website. If you want to serve HTTPS traffic […]
Dec 30th, 2022
https://aws.amazon.com/blogs/networking-and-content-delivery/hosting-internal-https-static-websites-with-alb-s3-and-privatelink/



Zac Nixon
  [8:36 AM](https://kubernetes.slack.com/archives/D09ARVB64TF/p1756136188696239)
Thanks for the link. I suspect that this works for your case because AFAIK, Private Link uses a static IP address.
[8:37](https://kubernetes.slack.com/archives/D09ARVB64TF/p1756136238044719)
Generally, we would like to supply more general solutions, this one has a pretty sharp edge where Target Group changes aren't propagated very fast.
```

TL;DR - There is some more work to be done here to accept this change. I don't really have the time until ~October to put more stuff on my plate.

**Comment by msvticket** [comment]
@zac-nixon I have now added support for regularly checking DNS (based on TTL). I have also removed any non-essential modifications. Please review!

File: pkg/backend/endpoint_resolver.go [hunk 5]
```diff
@@ -336,3 +382,173 @@ func convertCoreEndpointAddressToDiscoveryEndpoint(endpoint corev1.EndpointAddre
 	}
 	return ep
 }
+
+func (r *defaultEndpointResolver) watchExternalName(ctx context.Context, ttl uint32, svc *corev1.Service, port intstr.IntOrString) {
```
Comment (zac-nixon) [inline pkg/backend/endpoint_resolver.go]: This probably warrants some abstraction and tests.

File: pkg/backend/endpoint_resolver.go [hunk 6]
```diff
@@ -43,6 +52,8 @@ func NewDefaultEndpointResolver(k8sClient client.Client, podInfoRepo k8s.PodInfo
 		failOpenEnabled:      failOpenEnabled,
 		endpointSliceEnabled: endpointSliceEnabled,
 		logger:               logger,
+
+		externalNameReconcilers: make(map[string]context.CancelFunc),
```
Comment (zac-nixon) [inline pkg/backend/endpoint_resolver.go]: We'll need a mutex here. It's possible that there are threads competing to mutate this.

File: pkg/backend/endpoint_resolver.go [hunk 7]
```diff
@@ -336,3 +382,173 @@ func convertCoreEndpointAddressToDiscoveryEndpoint(endpoint corev1.EndpointAddre
...
+	case <-ctx.Done():
+		err = ctx.Err()
+	case _ = <-ticker.C:
+		var minTTL uint32
+		minTTL, err = r.reconcileExternalNameEndpointSlice(ctx, svc, port)
+		if minTTL == math.MaxUint32 || minTTL < 10 {
+			// Try again in 10 seconds if no valid TTL was found. Don't query in less than 10 seconds to avoid churn
+			minTTL = 10
+		}
+		go r.watchExternalName(ctx, minTTL, svc, port)
```
Comment (zac-nixon) [inline pkg/backend/endpoint_resolver.go]: IMO, we'll want to bound how many threads we spin up here. It would be great to reduce DNS load by preventing duplicate resolution threads if the same dns name is used.

**Comment by zac-nixon** [comment]
I am curious, did you consider just a pure in-memory cache instead of using endpoint slices?

**Comment by msvticket** [comment]
> I am curious, did you consider just a pure in-memory cache instead of using endpoint slices?

Yes, but that made the code way more complex.

File: pkg/backend/endpoint_resolver.go (hunk 6 above)
Comment (msvticket) [reply pkg/backend/endpoint_resolver.go]: OK

File: pkg/backend/endpoint_resolver.go (hunk 5 above)
Comment (msvticket) [reply pkg/backend/endpoint_resolver.go]: I have added tests

File: pkg/backend/endpoint_resolver.go (hunk 7 above)
Comment (msvticket) [reply pkg/backend/endpoint_resolver.go]: OK

**Comment by ksawerykarwacki** [comment]
Any chance that this will be merged ever? This PR is open for over two years, majority fo people just opted for dual ingress approach with alb pointing to nginx and nginx pointing to service. WIth EOL of nginx ingress controller next month, there will be no better time to introduce this feature.
